### PR TITLE
Feature spark create table as select and support keyed table dataframe api

### PR DESCRIPTION
**Pull request by hellojinsilei**
import spark create keyed table as select and replace CreateArcticTableStatement to CreateTableStatement

### CODE REVIEW

File: spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/execution/ExtendedArcticUnkeyedStrategy.scala [hunk 1]
```diff
@@ -29,17 +28,24 @@ import org.apache.spark.sql.connector.catalog.{Identifier, TableCatalog}
...
+    case CreateTableAsSelect(catalog, ident, parts, query, props, options, ifNotExists) =>
+      var propertiesMap: Map[String, String] = props
+      var optionsMap: Map[String, String] = options
+      if (options.contains("primary.keys")) {
+        propertiesMap += ("primary.keys" -> options("primary.keys"))
+        optionsMap += ("overwrite-mode" -> "dynamic")
+      }
+      val writeOptions = new CaseInsensitiveStringMap(JavaConverters.mapAsJavaMap(optionsMap))
+      CreateTableAsSelectExec(catalog, ident, parts, query, planLater(query),
+        propertiesMap, writeOptions, ifNotExists) :: Nil
```
Comment (baiyangtx) [inline spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/execution/ExtendedArcticUnkeyedStrategy.scala:48]: case CreateTableAsSelect() if isArcticTable && isHasPrimaryKey => 
    // code

only handle command run on **arctic  primary key** table

File: spark/v3.1/spark/src/main/scala/org/apache/spark/sql/catalyst/parser/ArcticCreateTablePrimaryKeyAstBuilder.scala [hunk 2]
```diff
@@ -18,63 +18,40 @@
...
+import org.apache.spark.unsafe.types.{CalendarInterval, UTF8String}
+import org.apache.spark.util.random.RandomSampler
 
 import java.util
 import java.util.Locale
+import javax.xml.bind.DatatypeConverter
 import scala.collection.JavaConverters._
+import scala.collection.mutable.ArrayBuffer
 
 class ArcticCreateTablePrimaryKeyAstBuilder(delegate: ParserInterface)
```
Comment (baiyangtx) [inline spark/v3.1/spark/src/main/scala/org/apache/spark/sql/catalyst/parser/ArcticCreateTablePrimaryKeyAstBuilder.scala:51]: change class name    ArcticCreateTablePrimaryKeyAstBuilder  =>  ArcticExtendSparkSqlAstBuilder

File: spark/v3.1/spark/src/test/java/com/netease/arctic/spark/SparkTestBase.java [hunk 3]
```diff
@@ -64,4 +69,21 @@ public void after() {
...
+        primaryKeysBlock = true;
+      } else if (StringUtils.startsWith(row[0].toString(), "# ") && primaryKeysBlock) {
+        primaryKeysBlock = false;
+      } else if (primaryKeysBlock){
+        descPrimaryKeys.add(row[0].toString());
+      }
+    }
+
+    Assert.assertEquals(primaryKeys.size(), descPrimaryKeys.size());
+    Assert.assertArrayEquals(primaryKeys.toArray(), descPrimaryKeys.toArray());
```
Comment (baiyangtx) [inline spark/v3.1/spark/src/test/java/com/netease/arctic/spark/SparkTestBase.java:87]: whatif mulit primary key column ?   assert equal using Set<>  and compare one by one

File: spark/v3.1/spark/src/test/java/com/netease/arctic/spark/TestCreateKeyedTableAsSelect.java [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+  @After
+  public void removeTables() {
+    sql("DROP TABLE IF EXISTS {0}.{1}", database, sourceTable);
+    sql("DROP TABLE IF EXISTS {0}.{1}", database, table);
+  }
+
+  @Test
+  public void testPrimaryKeyCTAS() {
+    sql("create table {0}.{1} using arctic primary key(id) AS SELECT * from {2}.{3}.{4}",
+        database, table, catalogName, database, sourceTable);
```
Comment (baiyangtx) [inline spark/v3.1/spark/src/test/java/com/netease/arctic/spark/TestCreateKeyedTableAsSelect.java:71]: can we support   ( create table primary key() using arctic  as select xxx ) 

make primary key define before  table provider ? 

whatis kudu sql grammar like ?

**Review by baiyangtx — COMMENTED** [review]
what is kudu grammer like ? 

primary key define before table provider or after that ?

File: spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/execution/ExtendedArcticUnkeyedStrategy.scala (hunk 1 above)
Comment (hellojinsilei) [reply spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/execution/ExtendedArcticUnkeyedStrategy.scala:48]: Could not identify whether it is an arctic table, `options.contains("primary.keys")` can prove ` isHasPrimaryKey`.
This code is compatible with the original CTAS

File: spark/v3.1/spark/src/test/java/com/netease/arctic/spark/TestCreateKeyedTableAsSelect.java [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+        expectedSpec, loadTable(identifier).spec());
+    assertEquals("Should have rows matching the source table",
+        sql("SELECT * FROM {0}.{1} ORDER BY id", database, table),
+        sql("SELECT * FROM {0}.{1} ORDER BY id", database, sourceTable));
+  }
+
+  @Test
+  public void testRTAS() {
+    sql("CREATE TABLE {0}.{1} USING arctic TBLPROPERTIES (''prop1''=''val1'', ''prop2''=''val2'')" +
+        "AS SELECT * FROM {2}.{3}", database, table, database, sourceTable);
```
Comment (baiyangtx) [inline spark/v3.1/spark/src/test/java/com/netease/arctic/spark/TestCreateKeyedTableAsSelect.java:117]: what is rtas ?  replace table as select ?

File: spark/v3.1/spark/src/test/java/com/netease/arctic/spark/TestCreateKeyedTableAsSelect.java [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+  }
+
+  @After
+  public void removeTables() {
+    sql("DROP TABLE IF EXISTS {0}.{1}", database, sourceTable);
+    sql("DROP TABLE IF EXISTS {0}.{1}", database, table);
+  }
+
+  @Test
+  public void testPrimaryKeyCTAS() {
```
Comment (baiyangtx) [inline spark/v3.1/spark/src/test/java/com/netease/arctic/spark/TestCreateKeyedTableAsSelect.java:69]: need a test has ctas with primary key and parition.

File: spark/v3.1/spark/src/test/java/com/netease/arctic/spark/TestCreateKeyedTableAsSelect.java (hunk 5 above)
Comment (baiyangtx) [reply spark/v3.1/spark/src/test/java/com/netease/arctic/spark/TestCreateKeyedTableAsSelect.java:117]: is this a test for ctas with out primary key and partition ?

**Review by baiyangtx — APPROVED** [review]

**Pull request merged**